src/libsyntax/parse/parser.rs

```diff
@@ -1777,7 +1777,25 @@ impl<'a> Parser<'a> {
                    require_name);
             let pat = self.parse_pat()?;
 
-            self.expect(&token::Colon)?;
+            if let Err(mut err) = self.expect(&token::Colon) {
+                // If we find a pattern followed by an identifier, it could be an (incorrect)
+                // C-style parameter declaration.
+                match self.parse_ident() {
```

Comment:
Can you check what happens if you write `fn foo(i32)`? I've got a feeling that as you're trying to parse an ident if it fails the parser is one step forward, which wold mean you are now standing past the `)`, causing knock on parse errors. You could use [`check_ident` + checking wether it is followed by a comma or a closing brace](https://github.com/rust-lang/rust/blob/79e82e054b75496d92012836d8c2b51c73fa9bd0/src/libsyntax/parse/parser.rs#L5188) instead. That way you have a slightly higher degree of confidence that you're emitting the suggestion in an appropriate time.

Reply:
The current code seems to work, so I've pushed a test case with it. I agree it'd be nicer to use check_ident, but wouldn't that remove the suggestion from the `fn one(i32 a b)` case?

Reply:
It would if you check for the following comma or brace, but what would `fn one(i32 a b)` be? At that point the code is so removed from the accepted syntax that all we can do is give up and let the user figure it out.